controllers/btpoperator_controller.go

```diff
@@ -1131,3 +1201,480 @@ func (r *BtpOperatorReconciler) watchConfigPredicates() predicate.Funcs {
+	}
+	(*u).SetManagedFields(nil)
+
+	for i, w := range webhooks {
+		webhookAsMap, ok := w.(map[string]interface{})
+		if !ok {
+			return fmt.Errorf("could not get webhookAsMap from unstructured")
+		}
+		clientConfigValue, ok := webhookAsMap[ClientConfigKey]
+		if !ok {
+			return fmt.Errorf("while geting client config in reconcileCaBundle")
+		}
+		clientConfigAsMap, ok := clientConfigValue.(map[string]interface{})
+		if !ok {
+			return fmt.Errorf("while casting client config in reconcileCaBundle")
+		}
+		caBundleValue, _ := clientConfigAsMap[CaBundleKey]
+		caBundleBytes, err := json.Marshal(caBundleValue)
+		if err != nil {
+			return err
+		}
+		if !bytes.Equal(caBundleBytes, expectedCaBundle) {
+			clientConfigAsMap[CaBundleKey] = expectedCaBundle
+			webhookAsMap[ClientConfigKey] = clientConfigAsMap
+			webhooks[i] = webhookAsMap
+		}
+	}
+
+	(*u).Object[WebhooksKey] = webhooks
+	return nil
+}
+
+func (r *BtpOperatorReconciler) reconcileWebhook(ctx context.Context, u **unstructured.Unstructured, expectedCa []byte, kind string) error {
+	if u == nil {
+		return fmt.Errorf("empty u param to reconcileCaBundle")
+	}
+
+	uu := &unstructured.Unstructured{}
+	uu.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   "admissionregistration.k8s.io",
+		Version: "v1",
+		Kind:    kind,
+	})
+	err := r.Get(ctx, client.ObjectKey{Name: (*u).GetName(), Namespace: (*u).GetNamespace()}, uu)
+	notFound := k8serrors.IsNotFound(err)
+	if err != nil && !notFound {
+		return err
+	}
+	if !notFound {
+		*u = uu
+	}
+	err = r.reconcileCaBundle(u, expectedCa)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *BtpOperatorReconciler) isWebhookSecretCertSignedByCaSecretCert(ctx context.Context) (bool, error) {
+	caCertificate, err := r.getCertificateFromSecret(ctx, CaSecret)
+	if err != nil {
+		return false, err
+	}
+
+	webhookCertificate, err := r.getCertificateFromSecret(ctx, WebhookSecret)
+	if err != nil {
+		return false, err
+	}
+
+	ok, err := certs.VerifyIfSecondCertificateIsSignedByFirstCertificate(caCertificate, webhookCertificate)
+	if err != nil {
+		return false, err
+	}
+
+	return ok, nil
+}
+
+func (r *BtpOperatorReconciler) doesCertificateExpireSoon(ctx context.Context, secretName string) (bool, error) {
+	secretData, err := r.getDataFromSecret(ctx, secretName)
+	if err != nil {
+		return false, err
+	}
+
+	dataKey, err := r.mapSecretNameToSecretDataKey(secretName)
+	if err != nil {
+		return false, err
+	}
+
+	certificate, err := r.getValueByKey(r.buildKeyNameWithExtension(dataKey, CertificatePostfix), secretData)
+	if err != nil {
+		return false, err
+	}
+	certificateDecoded, _ := pem.Decode(certificate)
+	certificateTemplate, err := x509.ParseCertificate(certificateDecoded.Bytes)
+	if err != nil {
+		return false, err
+	}
+
+	expirationTriggerBound := certificateTemplate.NotAfter.Add(ExpirationBoundary)
+	expiresSoon := time.Now().UTC().After(expirationTriggerBound)
+	return !expiresSoon, nil
```

Comment:
So `doesCertificateExpiresSoon` returns `!expiresSoon`? Weird.

Reply:
The meaning from my perspective is either:
- cert is expired or will expire soon
- we should regenerate it